boa/src/syntax/lexer/number.rs

```diff
@@ -0,0 +1,337 @@
+use crate::syntax::lexer::{token::Numeric, Token};
+use std::io::Read;
+use std::str::FromStr;
+
+/// Number literal lexing.
+///
+/// Assumes the digit is consumed by the cursor (stored in init).
+///
+/// More information:
+///  - [ECMAScript reference][spec]
+///  - [MDN documentation][mdn]
+///
+/// [spec]: https://tc39.es/ecma262/#sec-literals-numeric-literals
+/// [mdn]:
+#[derive(Debug, Clone, Copy)]
+pub(super) struct NumberLiteral {
+    init: char,
+    strict_mode: bool,
+}
+
+impl NumberLiteral {
+    /// Creates a new string literal lexer.
+    pub(super) fn new(init: char, strict_mode: bool) -> Self {
+        Self { init, strict_mode }
+    }
+}
+
+/// This is a helper structure
+///
+/// This structure helps with identifying what numerical type it is and what base is it.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum NumericKind {
+    Rational,
+    Integer(u8),
+    BigInt(u8),
+}
+
+impl NumericKind {
+    /// Get the base of the number kind.
+    fn base(self) -> u32 {
+        match self {
+            Self::Rational => 10,
+            Self::Integer(base) => base as u32,
+            Self::BigInt(base) => base as u32,
+        }
+    }
+
+    /// Converts `self` to BigInt kind.
+    fn to_bigint(self) -> Self {
+        match self {
+            Self::Rational => unreachable!("can not convert rational number to BigInt"),
+            Self::Integer(base) => Self::BigInt(base),
+            Self::BigInt(base) => Self::BigInt(base),
+        }
+    }
+}
+
+fn take_signed_integer<R>(
+    buf: &mut String,
+    cursor: &mut Cursor<R>,
+    kind: &NumericKind,
+) -> Result<(), Error>
+where
+    R: Read,
+{
+    // The next part must be SignedInteger.
+    // This is optionally a '+' or '-' followed by 1 or more DecimalDigits.
+    match cursor.next() {
+        Some(Ok('+')) => {
+            buf.push('+');
+            if !cursor.next_is_pred(&|c: char| c.is_digit(kind.base()))? {
+                // A digit must follow the + or - symbol.
+                return Err(Error::syntax("No digit found after + symbol"));
+            }
+        }
+        Some(Ok('-')) => {
+            buf.push('-');
+            if !cursor.next_is_pred(&|c: char| c.is_digit(kind.base()))? {
+                // A digit must follow the + or - symbol.
+                return Err(Error::syntax("No digit found after - symbol"));
+            }
+        }
+        Some(Ok(c)) if c.is_digit(kind.base()) => {
+            buf.push(c);
+        }
+        Some(Ok(c)) => {
+            return Err(Error::syntax(format!(
+                "When lexing exponential value found unexpected char: '{}'",
+                c
+            )));
+        }
+        Some(Err(e)) => {
+            return Err(e.into());
+        }
+        None => {
+            return Err(Error::syntax("No exponential value found"));
+        }
+    }
+
+    // Consume the decimal digits.
+    cursor.take_until_pred(buf, &|c: char| c.is_digit(kind.base()))?;
+
+    Ok(())
+}
```

Comment:
```suggestion
    match cursor.next().transpose()? {
        Some('+') => {
            buf.push('+');
            if !cursor.next_is_pred(&|c: char| c.is_digit(kind.base()))? {
                // A digit must follow the + or - symbol.
                return Err(Error::syntax("No digit found after + symbol"));
            }
        }
        Some('-') => {
            buf.push('-');
            if !cursor.next_is_pred(&|c: char| c.is_digit(kind.base()))? {
                // A digit must follow the + or - symbol.
                return Err(Error::syntax("No digit found after - symbol"));
            }
        }
        Some(c) if c.is_digit(kind.base()) => {
            buf.push(c);
        }
        Some(c) => {
            return Err(Error::syntax(format!(
                "When lexing exponential value found unexpected char: '{}'",
                c
            )));
        }
        None => {
            return Err(Error::syntax("No exponential value found"));
        }
    }

    // Consume the decimal digits.
    cursor.take_until_pred(buf, &|c: char| c.is_digit(kind.base()))?;

    Ok(())
}
```
We can use `.transpose()?` here to simplify the cases

Reply:
I wonder if I might just flip the cursor return to be Result<Option<>> - this might improve performance in most cases as transpose won't be needed

Reply:
> I wonder if I might just flip the cursor return to be `Result<Option<>>` - this might improve performance in most cases as transpose won't be needed

Yes. this would make the code much more cleaner

Reply:
> > I wonder if I might just flip the cursor return to be `Result<Option<>>` - this might improve performance in most cases as transpose won't be needed
> 
> Yes. this would make the code much more cleaner

The tradeoff comes from not being able to implement iterator

Reply:
If we don`t use other interator methods, I think it's  best if we implement `.next` returning `io::Result<Option<...>>` this would be much better.

Reply:
Ok I can go switch stuff around which should clean up the code a fair amount